wa/workloads/recentfling/__init__.py

```diff
@@ -0,0 +1,126 @@
+
+
+class Recentfling(Workload):
+
+    name = 'recentfling'
+    description = """
+    Tests UI jank on android devices.
+
+    For this workload to work, ``recentfling.sh`` and ``defs.sh`` must be placed
+    in ``~/.workload_automation/dependencies/recentfling/``. These can be found
+    in the `AOSP Git repository <https://android.googlesource.com/platform/system/extras/+/master/tests/workloads>`_.
+
+    To change the apps that are opened at the start of the workload you will need
+    to modify the ``defs.sh`` file. You will need to add your app to ``dfltAppList``
+    and then add a variable called ``{app_name}Activity`` with the name of the
+    activity to launch (where ``{add_name}`` is the name you put into ``dfltAppList``).
+
+    You can get a list of activities available on your device by running
+    ``adb shell pm list packages -f``
+    """
+    supported_platforms = ['android']
+
+    parameters = [
+        Parameter('loops', kind=int, default=3,
+                  description="The number of test iterations."),
+        Parameter('start_apps', kind=bool, default=True,
+                  description="""
+                  If set to ``False``,no apps will be started before flinging
+                  through the recent apps list (in which the assumption is
+                  there are already recently started apps in the list.
+                  """),
+        Parameter('device_name', kind=str,
+                  description="""
+                  If set, recentfling will use the fling parameters for this
+                  device instead of automatically guessing the device.  This can
+                  also be used if the device is not supported by recentfling,
+                  but its screensize is similar to that of one that is supported.
+
+                  For possible values, check your recentfling.sh.  At the time
+                  of writing, valid values are: 'shamu', 'hammerhead', 'angler',
+                  'ariel', 'mtp8996', 'bullhead' or 'volantis'.
+                  """),
+    ]
+
+    def initialise(self, context):  # pylint: disable=no-self-use
+        if context.target.get_sdk_version() < 23:
+            raise WorkloadError("This workload relies on ``dumpsys gfxinfo`` \
+                                 only present in Android M and onwards")
+
+    def setup(self, context):
+        self.defs_host = context.resolver.get(File(self, "defs.sh"))
+        self.defs_target = self.target.install(self.defs_host)
+        self.recentfling_host = context.resolver.get(File(self, "recentfling.sh"))
+        self.recentfling_target = self.target.install(self.recentfling_host)
+        self._kill_recentfling()
+        self.target.ensure_screen_is_on()
+
+    def run(self, context):
+        args = '-i {} '.format(self.loops)
+        if not self.start_apps:
+            args += '-N '
+        if self.device_name is not None:
+            args += '-d {}'.format(self.device_name)
```

Comment:
`device_name` does not have a default value and was never validated. Is this something that can be `initialize`'d from the target? (IIRC, this should be in the Android props).

Reply:
If you don't pass "-d" to `recentfling.sh`, it guesses the device name from the Android props.  The idea of `device_name` is that you can override this for devices that are not supported by recentfling but their screensize is similar to that of one that is supported.  See the `device_name` description for more info.  I've given it a default value of None.

Reply:
Ah, sorry, I missed that. Thanks for the explanation.